File: web/static/capture.js
/* The page work: remembering which browser this is, and stamping on the way out.
 *
 * The server that takes a note while the database is locked holds no key, so it
 * cannot tell a real note from one dropped into the spool by anything else able
 * to write the file. This is the other half of that: an enrolled browser keeps a
 * random key of its own and stamps every note it writes, and the *unlocked*
 * server checks the stamp against the copy inside the encrypted database.
 *
 * The bytes being stamped live in `stamp.js`, which is pure and tested against
 * the Python directly. This file only deals with the page.
 *
 * Nothing here is a security boundary on its own. A browser that cannot run it —
 * scripting off, storage blocked, an old engine — still writes notes; they are
 * simply held for review instead of filed, and the page says so at the time.
 */

(function () {
  "use strict";

  var DEVICE_ID = "counselog.device.id";
  var DEVICE_SECRET = "counselog.device.secret";
  var DEVICE_LABEL = "counselog.device.label";

  /* ── this browser's identity ─────────────────────────────────────────── */

  function readDevice() {
    try {
      var id = window.localStorage.getItem(DEVICE_ID);
      var secret = window.localStorage.getItem(DEVICE_SECRET);
      if (!id || !secret) return null;
      return { id: id, secret: secret,
               label: window.localStorage.getItem(DEVICE_LABEL) || id };
    } catch (error) {
      // Private browsing and blocked site data both throw here rather than
      // returning null. Not being enrolled is a supported state, so this is a
      // normal answer, not a failure.
      return null;
    }
  }

  function storeDevice(id, secret, label) {
    try {
      window.localStorage.setItem(DEVICE_ID, id);
      window.localStorage.setItem(DEVICE_SECRET, secret);
      window.localStorage.setItem(DEVICE_LABEL, label || id);
      return true;
    } catch (error) {
      return false;
    }
  }

  /* ── the capture form ────────────────────────────────────────────────── */

  function describeDevice(element, device) {
    if (!element) return;
    if (device) {
      element.textContent = "This browser writes as “" + device.label +
        "”. Notes written while locked are sealed until you sign in.";
      return;
    }
    element.textContent = element.dataset.unlocked === "1"
      ? "This browser is not enrolled, so notes it writes while locked will be " +
        "held for review. Enrol it on the Browsers page."
      : "This browser is not enrolled. It can still write; notes will be held " +
        "for review until you sign in and enrol it.";
  }

  function wireCaptureForm(form, stamping) {
    var state = document.getElementById("device-state");
    var device = readDevice();
    describeDevice(state, device);

    form.addEventListener("submit", function (event) {
      if (form.dataset.stamped === "yes") return;   // our own resubmit
      if (!device || !stamping || !window.crypto || !window.crypto.subtle) return;

      event.preventDefault();
      var text = stamping.normalizeNewlines(form.elements.text.value);
      var capturedAt = stamping.nowStamp();

      stamping.stamp(device.secret, text, capturedAt, device.id)
        .then(function (mac) {
          form.elements.device.value = device.id;
          form.elements.mac.value = mac;
          form.elements.captured_at.value = capturedAt;
        })
        .catch(function () {
          // Could not stamp. Send it anyway: a note held for review is far
          // better than a note the user believes they wrote and did not.
        })
        .then(function () {
          form.dataset.stamped = "yes";
          form.submit();
        });
    });
  }

  /* ── the enrolment page ──────────────────────────────────────────────── */

  function completeEnrolment(element) {
    var stored = storeDevice(element.dataset.deviceId,
                             element.dataset.deviceSecret,
                             element.dataset.deviceLabel);
    element.textContent = stored
      ? "This browser is now enrolled as “" + element.dataset.deviceLabel +
        "”. Notes it writes while locked will be filed, not held."
      : "This browser could not store its key, so the enrolment did not take. " +
        "Site data may be blocked here. You can revoke it below.";
    // The key was rendered into the page to get it here. Take it back out, so
    // it is not left sitting in the DOM.
    element.removeAttribute("data-device-secret");
  }

  document.addEventListener("DOMContentLoaded", function () {
    var enrolled = document.getElementById("enrolled");
    if (enrolled) completeEnrolment(enrolled);

    var form = document.getElementById("capture");
    if (form) wireCaptureForm(form, window.counselogStamp);
  });
})();

File: web/views/capture.py
"""Writing a note. The one thing that must work at any moment, in any state.

Two paths, and the split is exactly the security boundary:

  - **Signed in.** The key is here, so the note goes straight into the record.
    Nothing is gained by routing it through a sealed file first.
  - **Locked.** There is no key, so the note is sealed to the spool's public key
    and set aside. The browser stamps it on the way out, and the next sign-in
    decides whether to believe it.

The page is the same either way, and so is the box you type in. Which path a
note took is shown afterwards, because a note that will be held for review is
something the writer should hear about while they still remember writing it.
"""

from __future__ import annotations

from flask import (
    g,
    redirect,
    render_template,
    request,
    session as cookie,
    url_for,
)

from core import db, devices, intake, models, spool
from core.sanitize import normalize_newlines
from core.paths import notes_db_path, spool_db_path
from web.access import open_database, sessions_of
from web.sessions import NotSignedIn
from web.views.auth import take_intake_summary

WROTE_COOKIE = "wrote"
MAX_NOTE_CHARS = spool.MAX_NOTE_CHARS


def register(app) -> None:

    @app.get("/")
    def home():
        """The capture box, and nothing above it.

        This is the page a phone opens to. Anything that pushes the textarea
        below the fold makes the tool slower than a scrap of paper, which is the
        only competitor that matters.
        """
        return render_template(
            "capture.html",
            caller=g.caller,
            text="",
            error=None,
            wrote=cookie.pop(WROTE_COOKIE, None),
            intake_summary=take_intake_summary(),
            can_seal=_sealing_is_possible(),
            database_present=notes_db_path().exists(),
        )

    @app.post("/write")
    def write():
        # Normalised before anything else touches it. A form submission turns
        # every newline in a textarea into CRLF, and the browser stamped the
        # text as it held it — so without this, every multi-line note written
        # while locked would fail its check and be held for review.
        text = normalize_newlines(request.form.get("text") or "")

        def refuse(message: str, status: int = 400):
            return render_template(
                "capture.html", caller=g.caller, text=text, error=message,
                wrote=None, intake_summary={},
                can_seal=_sealing_is_possible(),
                database_present=notes_db_path().exists(),
            ), status

        if not text.strip():
            return refuse("Write something first.")
        if len(text) > MAX_NOTE_CHARS:
            return refuse("That note is longer than this can take. "
                          "Split it into a few notes.")

        try:
            dek = sessions_of().key(cookie.get("sid"))
        except NotSignedIn:
            dek = None

        if dek is not None:
            g.dek = dek
            try:
                outcome = _file_directly(text)
            except db.DatabaseError:
                # Signed in, but there is nothing to write into. Never echo the
                # underlying message: it carries a path.
                return refuse("There is no notes database on this machine yet. "
                              "Run `counselog init` first.", 503)
        else:
            outcome = _seal_for_later(text, request.form)
        if outcome is None:
            return refuse("This machine cannot hold a note written while locked "
                          "yet. Sign in once to set that up.", 503)

        # Redirect after posting so a refresh does not write the note twice.
        cookie[WROTE_COOKIE] = outcome
        return redirect(url_for("home"))


def _file_directly(text: str) -> dict:
    """Store the note in the record now, because the key is in hand."""
    conn = open_database()
    try:
        models.add_note(conn, text)
    finally:
        conn.close()
    return {"where": "record"}


def _seal_for_later(text: str, form) -> dict | None:
    """Seal the note to the spool. Returns None if this machine cannot yet.

    Everything the browser sends is treated as a claim, not a fact. A device id
    that is not one of ours, a stamp that is not the right shape, a timestamp
    that is not the one form we accept: none of these are refused, because
    refusing would throw away a note someone just wrote. They are normalised to
    something the drain will recognise as unstamped, and the writer is told the
    note will be held for review.
    """
    try:
        public = intake.published_key()
    except intake.IntakeError:
        return None

    device_id = form.get("device") or ""
    mac_hex = form.get("mac") or ""
    stamped = devices.is_device_id(device_id) and devices.is_mac_hex(mac_hex)

    captured_at = form.get("captured_at") or ""
    if not devices.is_captured_at(captured_at):
        # A browser that ran the script always sends the right shape, so this is
        # a browser that did not. Our clock is the honest answer, and the note
        # would be held for review anyway.
        captured_at = spool.utc_now()
        stamped = False

    conn = spool.connect(spool_db_path())
    try:
        spool.append(
            conn, public,
            text=text,
            captured_at=captured_at,
            device_id=device_id if stamped else devices.UNSTAMPED,
            mac=bytes.fromhex(mac_hex) if stamped else b"\x00" * 32,
        )
    finally:
        conn.close()
    return {"where": "spool", "stamped": stamped}


def _sealing_is_possible() -> bool:
    """Can a note be written while locked on this machine yet?

    False on a fresh install until someone signs in once, because the keypair is
    made with the database open. The page says so rather than presenting a box
    that would fail on submit (Guideline 2).
    """
    try:
        intake.published_key()
    except intake.IntakeError:
        return False
    return True


__all__ = ["register", "MAX_NOTE_CHARS"]
